Mathlib/NumberTheory/Padics/PadicVal.lean

```diff
@@ -0,0 +1,542 @@
+#align padic_val_rat_of_nat padicValRat_of_nat
+
+/-- A simplification of `padic_val_nat` when one input is prime, by analogy with
+`padic_val_rat_def`. -/
+theorem padicValNat_def [hp : Fact p.Prime] {n : ℕ} (hn : 0 < n) :
+    padicValNat p n = (multiplicity p n).get (multiplicity.finite_nat_iff.2 ⟨hp.out.ne_one, hn⟩) :=
+  dif_pos ⟨hp.out.ne_one, hn⟩
+#align padic_val_nat_def padicValNat_def
+
+theorem padicValNat_def' {n : ℕ} (hp : p ≠ 1) (hn : 0 < n) :
+    ↑(padicValNat p n) = multiplicity p n := by simp [padicValNat, hp, hn]
+#align padic_val_nat_def' padicValNat_def'
+
+@[simp]
+theorem padicValNat_self [Fact p.Prime] : padicValNat p p = 1 := by
+  rw [padicValNat_def (@Fact.out p.Prime).pos]
+  simp
+#align padic_val_nat_self padicValNat_self
+
+theorem one_le_padicValNat_of_dvd {n : ℕ} [hp : Fact p.Prime] (hn : 0 < n) (div : p ∣ n) :
+    1 ≤ padicValNat p n := by
+  rwa [← PartENat.coe_le_coe, padicValNat_def' hp.out.ne_one hn, ← pow_dvd_iff_le_multiplicity,
+    pow_one]
+#align one_le_padic_val_nat_of_dvd one_le_padicValNat_of_dvd
+
+theorem dvd_iff_padicValNat_ne_zero {p n : ℕ} [Fact p.Prime] (hn0 : n ≠ 0) :
+    p ∣ n ↔ padicValNat p n ≠ 0 :=
+  ⟨fun h => one_le_iff_ne_zero.mp (one_le_padicValNat_of_dvd hn0.bot_lt h), fun h =>
+    Classical.not_not.1 (mt padicValNat.eq_zero_of_not_dvd h)⟩
+#align dvd_iff_padic_val_nat_ne_zero dvd_iff_padicValNat_ne_zero
+
+end padicValNat
+
+namespace padicValRat
+
+open multiplicity
+
+variable {p : ℕ} [hp : Fact p.Prime]
+
+/-- The multiplicity of `p : ℕ` in `a : ℤ` is finite exactly when `a ≠ 0`. -/
+theorem finite_int_prime_iff {a : ℤ} : Finite (p : ℤ) a ↔ a ≠ 0 := by
+  simp [finite_int_iff, hp.1.ne_one]
+#align padic_val_rat.finite_int_prime_iff padicValRat.finite_int_prime_iff
+
+/-- A rewrite lemma for `padic_val_rat p q` when `q` is expressed in terms of `rat.mk`. -/
+protected theorem defn (p : ℕ) [hp : Fact p.Prime] {q : ℚ} {n d : ℤ} (hqz : q ≠ 0)
+    (qdf : q = n /. d) :
+    padicValRat p q =
+      (multiplicity (p : ℤ) n).get
+          (finite_int_iff.2 ⟨hp.1.ne_one, fun hn => by simp_all⟩) -
+        (multiplicity (p : ℤ) d).get
+          (finite_int_iff.2 ⟨hp.1.ne_one, fun hd => by simp_all⟩) := by
+  have hd : d ≠ 0 := Rat.mk_denom_ne_zero_of_ne_zero hqz qdf
+  let ⟨c, hc1, hc2⟩ := Rat.num_den_mk hd qdf
+  rw [padicValRat.multiplicity_sub_multiplicity hp.1.ne_one hqz]
+  simp only [Nat.isUnit_iff, hc1, hc2]
+  rw [multiplicity.mul' (Nat.prime_iff_prime_int.1 hp.1),
+    multiplicity.mul' (Nat.prime_iff_prime_int.1 hp.1)]
+  rw [Nat.cast_add, Nat.cast_add]
+  simp_rw [Int.coe_nat_multiplicity p q.den]
+  ring
+  -- Porting note: was
+  -- simp only [hc1, hc2, multiplicity.mul' (Nat.prime_iff_prime_int.1 hp.1),
+  --   hp.1.ne_one, hqz, pos_iff_ne_zero, Int.coe_nat_multiplicity p q.den
+#align padic_val_rat.defn padicValRat.defn
+
+/-- A rewrite lemma for `padic_val_rat p (q * r)` with conditions `q ≠ 0`, `r ≠ 0`. -/
```

Comment:
```suggestion
/-- A rewrite lemma for `padicValRat p (q * r)` with conditions `q ≠ 0`, `r ≠ 0`. -/
```
Could you fix that name and others everywhere in the file?

Reply:
All done.